xray/services/scan.go

```diff
@@ -0,0 +1,210 @@
+package services
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/jfrog/jfrog-client-go/artifactory/services/utils"
+	"github.com/jfrog/jfrog-client-go/auth"
+	"github.com/jfrog/jfrog-client-go/http/jfroghttpclient"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+)
+
+const (
+	scanGraphAPI             = "api/v1/scan/graph"
+	repoPathQueryParam       = "?repo_path="
+	projectQueryParam        = "?project="
+	defaultMaxWaitMinutes    = 5 * time.Minute // 5 minutes
+	defaultSyncSleepInterval = 5               // 5 seconds
+)
+
+type ScanService struct {
+	client         *jfroghttpclient.JfrogHttpClient
+	XrayDetails    auth.ServiceDetails
+	MaxWaitMinutes time.Duration
+}
+
+// NewScanService creates a new service to scan Binaries and Sources.
+func NewScanService(client *jfroghttpclient.JfrogHttpClient) *ScanService {
+	return &ScanService{client: client}
+}
+
+func (ss *ScanService) ScanGraph(scanParams XrayGraphScanParams) (string, error) {
+	httpClientsDetails := ss.XrayDetails.CreateHttpClientDetails()
+	utils.SetContentType("application/json", &httpClientsDetails.Headers)
+	requestBody, err := json.Marshal(scanParams.Graph)
+	if err != nil {
+		return "", errorutils.CheckError(err)
+	}
+	url := ss.XrayDetails.GetUrl() + scanGraphAPI
+	if scanParams.ProjectKey != "" {
+		url += projectQueryParam + scanParams.ProjectKey
+	} else if scanParams.RepoPath != "" {
+		url += repoPathQueryParam + scanParams.RepoPath
+	}
+	resp, body, err := ss.client.SendPost(url, requestBody, &httpClientsDetails)
+	if err != nil {
+		return "", err
+	}
+
+	if err = errorutils.CheckResponseStatus(resp, http.StatusOK, http.StatusCreated); err != nil {
+		return "", errorutils.CheckError(errors.New("Server Response: " + resp.Status))
+	}
+	scanResponse := RequestScanResponse{}
+	if err = json.Unmarshal(body, &scanResponse); err != nil {
+		return "", errorutils.CheckError(err)
+	}
+	return scanResponse.ScanId, err
+}
+
+func (ss *ScanService) GetScanGraphResults(scanId string) (*ScanResponse, error) {
+	maxWaitMinutes := defaultMaxWaitMinutes
+	if ss.MaxWaitMinutes > 0 {
+		maxWaitMinutes = ss.MaxWaitMinutes
+	}
+	httpClientsDetails := ss.XrayDetails.CreateHttpClientDetails()
+	utils.SetContentType("application/json", &httpClientsDetails.Headers)
+
+	message := fmt.Sprintf("Sync: Get Scan Graph Results. Scan ID:%s...", scanId)
+	// The scan request may take time to be calculated, we expect to receive a 202 response until the compleation of the requested scan.
```

Comment:
The scan request may take time to be calculated, we expect to receive a 202 response until the compleation of the requested scan.
-->
The scan request may take some time to complete. We expect to receive a 202 response, until the completion.